From 6e249fd1743e5b0f126ab221b2a431cf40c92625 Mon Sep 17 00:00:00 2001
From: Jason Lowe <jlowe@yahoo-inc.com>
Date: Thu, 25 May 2017 16:07:52 -0500
Subject: [PATCH 2501/2848] YARN-6643. TestRMFailover fails rarely due to port
 conflict. Contributed by Robert Kanter

(cherry picked from commit 3fd6a2da4e537423d1462238e10cc9e1f698d1c2)
(cherry picked from commit 4c44ff69df4979d40dc78a0dbd00de967627643a)

Change-Id: I073bd2693f1cd34888a62c521781944726895887
---
 .../yarn/server/resourcemanager/HATestUtil.java    |    9 ++++++---
 .../hadoop/yarn/server/TestMiniYarnCluster.java    |    4 +++-
 2 files changed, 9 insertions(+), 4 deletions(-)

diff --git a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/test/java/org/apache/hadoop/yarn/server/resourcemanager/HATestUtil.java b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/test/java/org/apache/hadoop/yarn/server/resourcemanager/HATestUtil.java
index 710ce87..ac245c3 100644
--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/test/java/org/apache/hadoop/yarn/server/resourcemanager/HATestUtil.java
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/test/java/org/apache/hadoop/yarn/server/resourcemanager/HATestUtil.java
@@ -18,16 +18,19 @@
 package org.apache.hadoop.yarn.server.resourcemanager;
 
 import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.net.ServerSocketUtil;
 import org.apache.hadoop.yarn.conf.HAUtil;
 import org.apache.hadoop.yarn.conf.YarnConfiguration;
 
+import java.io.IOException;
+
 public class HATestUtil {
 
   public static void setRpcAddressForRM(String rmId, int base,
-      Configuration conf) {
+      Configuration conf) throws IOException {
     for (String confKey : YarnConfiguration.getServiceAddressConfKeys(conf)) {
-      setConfForRM(rmId, confKey, "0.0.0.0:" + (base +
-          YarnConfiguration.getRMDefaultPortNumber(confKey, conf)), conf);
+      setConfForRM(rmId, confKey, "0.0.0.0:" + ServerSocketUtil.getPort(base +
+          YarnConfiguration.getRMDefaultPortNumber(confKey, conf), 10), conf);
     }
   }
 
diff --git a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-tests/src/test/java/org/apache/hadoop/yarn/server/TestMiniYarnCluster.java b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-tests/src/test/java/org/apache/hadoop/yarn/server/TestMiniYarnCluster.java
index 3e35bd0..fbdedc2 100644
--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-tests/src/test/java/org/apache/hadoop/yarn/server/TestMiniYarnCluster.java
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-tests/src/test/java/org/apache/hadoop/yarn/server/TestMiniYarnCluster.java
@@ -25,6 +25,8 @@
 import org.junit.Assert;
 import org.junit.Test;
 
+import java.io.IOException;
+
 public class TestMiniYarnCluster {
   @Test
   public void testTimelineServiceStartInMiniCluster() throws Exception {
@@ -116,7 +118,7 @@ public void testTimelineServiceStartInMiniCluster() throws Exception {
   }
 
   @Test
-  public void testMultiRMConf() {
+  public void testMultiRMConf() throws IOException {
     String RM1_NODE_ID = "rm1", RM2_NODE_ID = "rm2";
     int RM1_PORT_BASE = 10000, RM2_PORT_BASE = 20000;
     Configuration conf = new YarnConfiguration();
-- 
1.7.9.5

